src/run.rs

```diff
@@ -0,0 +1,171 @@
+
+    // build an ISA for the current machine
+    host_isa_builder()
+        .expect("A builder could not be created")
+        .finish(flags)
+}
+
+/// Iterate over all of the files passed as arguments. TODO eventually this should recurse into
+/// subdirectories, ignore non-CLIF files, ignore hidden files, etc.
+fn iterate_files(files: Vec<String>) -> Box<dyn Iterator<Item = PathBuf>> {
+    let iterator = files.into_iter().map(PathBuf::from);
+    Box::new(iterator)
+}
+
+/// Determine a path's file name
+fn get_file_name(path: &PathBuf) -> Result<String, String> {
+    let os_str = path
+        .file_name()
+        .ok_or(String::from("Unable to figure out the file name"))?;
+    Ok(String::from(os_str.to_string_lossy()))
+}
+
+/// Run all functions in a file that are succeeded by "run:" comments
+fn run_single_file(path: &PathBuf, isa: &dyn TargetIsa) -> Result<(), String> {
+    let file_contents = read_to_string(&path).map_err(|e| e.to_string())?;
+    run_file_contents(file_contents, isa)
+}
+
+/// Main body of `run_single_file` separated for testing
+fn run_file_contents(file_contents: String, isa: &dyn TargetIsa) -> Result<(), String> {
+    let test_file = parse_test(&file_contents, None, None).map_err(|e| e.to_string())?;
+    for (func, Details { comments, .. }) in test_file.functions {
+        // TODO this may need to iterate over the comments
+        if comments.iter().filter(|c| c.text.contains("run:")).count() > 0 {
+            run_single_function(func, isa)?
+        }
+    }
+    Ok(())
+}
+
+/// Compile and execute a single function, expecting a boolean to be returned; a 'true' value is
+/// interpreted as a successful test execution and mapped to Ok whereas a 'false' value is
+/// interpreted as a failed test and mapped to Err.
+fn run_single_function(func: Function, isa: &dyn TargetIsa) -> Result<(), String> {
+    if !(func.signature.params.is_empty()
+        && func.signature.returns.len() == 1
+        && func.signature.returns.first().unwrap().value_type.is_bool())
+    {
+        return Err(String::from(
+            "Functions must have a signature like: () -> boolean",
+        ));
+    }
+    // TODO observed segfaults with return types of b64
+
+    // set up the context
+    let mut context = Context::new();
+    context.func = func;
+
+    // compile and encode the result to machine code
+    let mut relocs = NullRelocSink {};
+    let mut traps = NullTrapSink {};
+    let mut mem = vec![];
+    context
+        .compile_and_emit(isa, &mut mem, &mut relocs, &mut traps)
+        .map_err(|err| pretty_error(&context.func, Some(isa), err))?;
+
+    // make the machine code executable in Rust
+    let callable_fn: fn() -> bool = unsafe {
+        protect(mem.as_mut_ptr(), mem.len(), Protection::ReadWriteExecute).expect("..."); // this segfaults if it is only ReadExecute, a bactrace in `crashdumpctl debug` shows the explosion during deallocation at the end of this block
```

Comment:
You should `mmap` pages, instead of `mprotect`ing the heap. The later is UB. Also why not use `cranelift-simplejit`?

Reply:
It wasn't exactly clear to me how to take a CLIF file and compile it with SimpleJIT; most of the clif-util code seems to interact directly with `Context` so I had some examples there. Also, in talking to @nbp he pointed me to his [HolyJIT](https://github.com/nbp/holyjit) project which was another helpful example that directly compiles using `Context`. In that project he does `mmap` in some memory to receive the emitted bytes so I can follow what he did there.

Before I continue, though, is there some way I can use the pre-existing SimpleJIT to do what I'm doing now with the CLIF test files?

Reply:
@bjorn3 ok, I think the should be resolved in the latest push.